src/core/shape_inference/include/grid_sample_shape_inference.hpp

```diff
@@ -18,19 +18,43 @@ void shape_infer(const GridSample* op, const std::vector<T>& input_shapes, std::
                           input_shapes.size() == 2 && output_shapes.size() == 1,
                           "Incorrect number of input/output shapes in GridSample's shape inference function");
     const auto& data_shape = input_shapes[0];
-    NODE_VALIDATION_CHECK(op, data_shape.rank().same_scheme(4), "The supported shape of the input data tensor is 4D.");
+    NODE_VALIDATION_CHECK(op, data_shape.rank().compatible(4), "The supported shape of the input data tensor is 4D.");
     const auto& grid_shape = input_shapes[1];
-    NODE_VALIDATION_CHECK(op, grid_shape.rank().same_scheme(4), "The supported shape of the grid tensor is 4D.");
-    NODE_VALIDATION_CHECK(op,
-                          grid_shape[3].same_scheme(2),
-                          "The last dimension of grid tensor's shape has to be equal to 2.");
+    NODE_VALIDATION_CHECK(op, grid_shape.rank().compatible(4), "The supported shape of the grid tensor is 4D.");
 
-    NODE_VALIDATION_CHECK(op,
-                          data_shape[0].same_scheme(grid_shape[0]),
-                          "The batch dimension in the input data tensor's shape doesn't match the batch dimension in "
-                          "the grid tensor's shape.");
+    Dimension batch_dim = Dimension::dynamic();
+    Dimension channel_dim = Dimension::dynamic();
+    Dimension H_out_dim = Dimension::dynamic();
+    Dimension W_out_dim = Dimension::dynamic();
+
+    if (grid_shape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(op,
+                              grid_shape[3].compatible(2),
+                              "The last dimension of grid tensor's shape has to be equal to 2.");
+        batch_dim = grid_shape[0];
+        H_out_dim = grid_shape[1];
+        W_out_dim = grid_shape[2];
```

Comment:
Conversion from `Dimension` to `StaticDimension` will fail for CPU:
https://github.com/openvinotoolkit/openvino/blob/bb3868d8cd7a0b9710a2148986ce6a8323a3a16e/src/plugins/intel_cpu/src/utils/shape_inference/static_dimension.hpp#L33-L35

Reply:
You can create `T output_shape(4);` and overwrite dimensions in this vector instead of separate variable for each dim.

Reply:
Good catch thanks. This constructor is not available though but there's a way to create 4 default constructed dimensions generically using the `resize()` method.